### PR TITLE
Python spackbot

### PR DESCRIPTION
**Pull request by tgamblin**
This is a set of commits to convert from Ruby to Python.  It's still WIP for now.

### CODE REVIEW

File: .gitignore
Comment (adamjstewart) [inline .gitignore:1]: Let's just replace this with https://github.com/github/gitignore/blob/master/Python.gitignore

File: README.md
Comment (adamjstewart) [inline README.md:9]:
```suggestion
* automatically labeling pull requests based on the files they modify
```

File: pyproject.toml
Comment (adamjstewart) [inline pyproject.toml:3]: Are we planning on "releases" for spack-bot? I was assuming this would be updated live any time a commit was merged.

File: spackbot/pull_requests.py
Comment (adamjstewart) [inline spackbot/pull_requests.py:92]: If I'm understanding this correctly, this will comment on PRs and ask maintainers to review when the PR is reopened or closed. I think we only want to comment on PRs once when a PR is first opened, otherwise we'll be commenting multiple times on the same PR.

File: spackbot/pull_requests.py
Comment (adamjstewart) [inline spackbot/pull_requests.py:66]: Is this left over from Ruby?

**Review by adamjstewart — COMMENTED** [review]
Looks like there is still some Ruby code that needs to be converted to Python. And the labeling logic needs to be ported. And the actual code to comment on PRs.

File: spackbot/pull_requests.py
Comment (tgamblin) [reply spackbot/pull_requests.py:66]: yep

File: spackbot/pull_requests.py
Comment (tgamblin) [reply spackbot/pull_requests.py]: this was an artifact of testing

**Comment by tgamblin** [comment]
@adamjstewart: `pr_add_labels.py` is checked in and works; `pr_add_reviewers.py` coming next, and then this will be done.

I think the python ends up being clearer than the ruby code in the end (`gidgethub` is pretty nice).

**Comment by tgamblin** [comment]
@adamjstewart this is ready for another review -- seems like it's the first working version of the Python spackbot.

File: spackbot/pr_add_labels.py
Comment (adamjstewart) [inline spackbot/pr_add_labels.py:25]: I would indent this to the same level as the code, at first I thought this was the entire dict and not just an entry in the dict. Prob won't be an issue in the docs (are we making docs??) but a little difficult to read in comment form.

Comment (adamjstewart) [reply spackbot/pr_add_labels.py:25]: I wonder if we actually need support for lists since regexes can express that

File: spackbot/pr_add_labels.py
Comment (adamjstewart) [inline spackbot/pr_add_labels.py:94]: Should we add `lib/spack/llnl` to this?

File: spackbot/pr_add_labels.py
Comment (adamjstewart) [inline spackbot/pr_add_labels.py:84]: Should one of these be deleted?

File: spackbot/pr_add_labels.py
Comment (adamjstewart) [inline spackbot/pr_add_labels.py:115]: Not entirely sure if this is still needed now that sbang has its own separate repo, but won't hurt

File: spackbot/pr_add_labels.py
Comment (adamjstewart) [inline spackbot/pr_add_labels.py:117]: I would also include csh and fish

File: spackbot/pr_add_labels.py
Comment (adamjstewart) [inline spackbot/pr_add_labels.py:153]: One of the ways I can tell if a section of code was written by you is if the comments are lowercase lol. According to [PEP-8](https://www.python.org/dev/peps/pep-0008/#comments):

> Comments should be complete sentences. The first word should be capitalized, unless it is an identifier that begins with a lower case letter (never alter the case of identifiers!).

Although I usually don't follow this rule for inline comments, only block comments. This is one of my many OCD things 😄

File: spackbot/pr_add_labels.py
Comment (adamjstewart) [inline spackbot/pr_add_labels.py:133]: PEP-8 seems to suggest that docstrings shouldn't start with a space: https://www.python.org/dev/peps/pep-0008/#documentation-strings

File: spackbot/pr_add_labels.py
Comment (adamjstewart) [inline spackbot/pr_add_labels.py:164]: Let's just take a moment to stop and appreciate the fact that this script doesn't have to support Python 2.6 and can use all the niceties of modern Python

File: spackbot/pr_add_reviewers.py
Comment (adamjstewart) [inline spackbot/pr_add_reviewers.py:44]: Someday we may want to refactor things and move this to a separate file so multiple bot functions can use it, but let's way until we actually need to. "Premature optimization is the root of all evil" is the only thing I've learned in grad school...

File: spackbot/pr_add_reviewers.py
Comment (adamjstewart) [inline spackbot/pr_add_reviewers.py:58]: Can remove

File: spackbot/pr_add_reviewers.py
Comment (adamjstewart) [inline spackbot/pr_add_reviewers.py:85]: This could use a better docstring with type annotations and things. Actually, now that we don't have to support old Python, how do you feel about mypy? I've been using it for all of my projects. It hasn't yet caught much when my code is already covered by unit tests, but since we don't have unit tests (yet) it may actually help prevent bugs.

File: spackbot/pr_add_reviewers.py
Comment (adamjstewart) [inline spackbot/pr_add_reviewers.py:138]: All this async and await stuff is really fancy. Someday I would like to learn how to write code like this.

File: spackbot/pr_add_reviewers.py
Comment (adamjstewart) [inline spackbot/pr_add_reviewers.py:227]: Should this be dropped before merging? We want to add labels on synch but not re-ping the same maintainers.

File: spackbot/pr_add_reviewers.py
Comment (adamjstewart) [inline spackbot/pr_add_reviewers.py:234]: Probably can be removed, we should use logging for all logging

**Review by adamjstewart — COMMENTED** [review]
Gave things an initial review. We should also get someone else to review before deploying live. It would be good to set up a test bed to make sure we have all the permissions right before making everything live.

File: spackbot/pr_add_labels.py
Comment (tgamblin) [reply spackbot/pr_add_labels.py:25]: makes it easier to parse for mere mortals.

File: spackbot/pr_add_labels.py
Comment (tgamblin) [reply spackbot/pr_add_labels.py:94]: done

File: spackbot/pr_add_labels.py
Comment (tgamblin) [reply spackbot/pr_add_labels.py:84]: deleted

File: spackbot/pr_add_labels.py
Comment (tgamblin) [reply spackbot/pr_add_labels.py:115]: we still need this, I think, as we copy it into the spack repo.

File: spackbot/pr_add_labels.py
Comment (tgamblin) [reply spackbot/pr_add_labels.py:117]: done

File: spackbot/pr_add_labels.py
Comment (tgamblin) [reply spackbot/pr_add_labels.py:153]: I think I started lowercasing more based on your guidelines for command line option help years ago.  I can fix it.

File: spackbot/pr_add_reviewers.py
Comment (tgamblin) [reply spackbot/pr_add_reviewers.py:85]: can we hold off on the mypy and maybe add that later?

Comment (tgamblin) [reply spackbot/pr_add_reviewers.py:85]: I have mixed feelings about it.  I've seen a bunch of people struggle to understand how to do the type annotations, though that may be because in Spack they have to be in comments.

File: spackbot/pr_add_reviewers.py
Comment (tgamblin) [reply spackbot/pr_add_reviewers.py:138]: It wasn't that hard to get used to -- it's just lightweight threads.

File: spackbot/pr_add_reviewers.py
Comment (tgamblin) [reply spackbot/pr_add_reviewers.py:227]: So the thing here is that they may modify more files in the meantime.  What do you want to do about that?

Comment (adamjstewart) [reply spackbot/pr_add_reviewers.py:227]: My thoughts:

1. If we can request review instead of comment on PR, then synchronize is fine
2. If we can only comment on PR, we don't want to comment after every single commit

File: spackbot/pr_add_reviewers.py
Comment (adamjstewart) [reply spackbot/pr_add_reviewers.py:85]: We can def hold off on that

File: spackbot/pr_add_reviewers.py
Comment (tgamblin) [reply spackbot/pr_add_reviewers.py:227]: ok, how about we remove it for now, and add that functionality later.

Comment (tgamblin) [reply spackbot/pr_add_reviewers.py:227]: I have this feeling we'll need to iterate anyway so I just want to get this thing live.

File: pyproject.toml
Comment (tgamblin) [reply pyproject.toml:3]: Wouldn't mind doing this eventually but I don't think it's completely necessary for attempt number 1.

File: spackbot/pr_add_reviewers.py
Comment (adamjstewart) [reply spackbot/pr_add_reviewers.py]: Completely agree

**Comment by tgamblin** [comment]
Ok I think this is pretty good for now -- can you merge when ready?

**Comment by tgamblin** [comment]
One nice thing about GitHub apps is that we can deploy this at a URL and we do not have to install on Spack initially.  What I would like to do is set up automated builds with Dockerhub, though.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
